src/main/java/com/github/karmazyn/locking/LockerImpl.java

```diff
@@ -0,0 +1,65 @@
+package com.github.karmazyn.locking;
+
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.locks.ReentrantLock;
+
+public class LockerImpl implements Locker {
+
+    private final ConcurrentHashMap<Object, MutexImpl> locks = new ConcurrentHashMap<>();
+
+    @Override
+    public Mutex lock(Object object) {
+
+        if (object == null) {
+            throw new IllegalArgumentException();
+        }
+
+        boolean locked = false;
+        MutexImpl mutex = null;
+
+        while (!locked) {
+            mutex = locks.computeIfAbsent(object, (key) -> new MutexImpl(key));
+            locked = mutex.lockInternal();
+        }
+
+        return mutex;
+    }
+
+    private boolean remove(Object key, Object expected) {
+        return locks.remove(key, expected);
+    }
+
+    private class MutexImpl implements Mutex {
+
+        private final Object key;
+        private final ReentrantLock internalLock;
+        private volatile boolean isRemoved = false;
+
+        MutexImpl(Object key) {
+            this.internalLock = new ReentrantLock();
+            this.key = key;
+        }
+
+        private boolean lockInternal() {
+            internalLock.lock();
+            if (isRemoved) {
+                internalLock.unlock();
+                return false;
+            }
+            return true;
+        }
+
+        @Override
+        public void release() {
```

Comment:
good at least once it is released and there are no more locked threads, this mutex is as good as dead, client cannot use it again

Reply:
Yes, you are right. The reference to the mutex is useless, but it can acquire the lock again using Locker.  Adding lock method to Mutex would probably mean removing the locking part from Locker (it would just serve as registrar/lookup).  It is a good idea, since it would make the Mutex more coherent with Lock interface. I've considered implementing it that way, but it is quite a different concept to what was in the requirements. 
Since I didn't know the use cases where this would be used, I've decided to stick to what was in the requirements.